core/testing.py

```diff
@@ -1659,3 +1660,45 @@ def pytest_configure(config):
         "markers", "elasticsearch: mark test as requiring elasticsearch"
     )
     config.addinivalue_line("markers", "minio: mark test as requiring minio")
+
+
+class MockDeliveryMechanism:
```

Comment:
Sounds good. I'd prefer that too!

Reply:
Sadly, it turns out that it buys almost nothing in this case, and does actually make things harder to understand. The problem is that we need to override `__eq__` and `__repr__` and `__str__`: The first is needed in order to compare the resulting list of delivery mechanisms in the tests, and the latter two are needed so that the "these two lists aren't equal" message you get when a test fails is actually readable. With plain mocks, you'll get this:

```
E         Full diff:
E           [
E         -  <MagicMock spec='LicensePoolDeliveryMechanism' id='139679558147664'>,
E         -  <MagicMock spec='LicensePoolDeliveryMechanism' id='139679544032080'>,
E         -  <MagicMock spec='LicensePoolDeliveryMechanism' id='139679543684352'>,
E         -  <MagicMock spec='LicensePoolDeliveryMechanism' id='139679558309872'>,
E         ?                                                              ^ - ^^
E         +  <MagicMock spec='LicensePoolDeliveryMechanism' id='139679558270880'>,
E         ?                                                              ^^  ^^
E         -  <MagicMock spec='LicensePoolDeliveryMechanism' id='139679558310064'>,
E         -  <MagicMock spec='LicensePoolDeliveryMechanism' id='139679558147616'>,
```

... which is amazingly unhelpful. After assigning each of the magic methods manually to get good test output... You almost have the original code in its entirety. :disappointed:

Reply:
I think the mock library will be more useful when dealing with things that are inherently effectful. The basic problem here is that the library delivery mechanisms should have been simple immutable value classes and not the pile of side effects and database accesses that they actually are now. Extracting the parts that we actually care about for this code into a simple explicit mock gives shorter code in this case.

Reply:
Yep, don't think I wasn't eyeing the potential type errors. :sweat_smile: 

Unfortunately, using the actual class brings too much baggage. It becomes necessary to set up a lot of database state to get something that's conceptually just a small list of tuples.

I'm going to keep hammering at the mock version, but just be aware that it's not going to be smaller than the original!

Reply:
Managed to shave some corners off!

Reply:
This looks good to me. One final suggestion would be to make the static `mechanism_of` function into a `pytest.fixture`. Some day I'd like to get our tests to the point where we have a set of pytest.fixtures that can be shared across all the tests. Adding some useful fixtures as we add new tests, gets us moving in that direction.  Plus it is the pytest recommended way to return testing functions and fixtures like this. 

``` python
class TestFormatPriorities:
    @pytest.fixture
    def mock_delivery(self):
        def delivery_mechanism(
            drm_scheme: Optional[str] = None,
            content_type: Optional[str] = "application/epub+zip",
        ) -> DeliveryMechanism:
            def _delivery_eq(self, other):
                return (
                    self.drm_scheme == other.drm_scheme
                    and self.content_type == other.content_type
                )

            def _delivery_repr(self):
                return f"DeliveryMechanism(drm_scheme={self.drm_scheme}, content_type={self.content_type})"

            _delivery = MagicMock(spec=DeliveryMechanism)
            _delivery.drm_scheme = drm_scheme
            _delivery.content_type = content_type
            _delivery.__eq__ = _delivery_eq
            _delivery.__repr__ = _delivery_repr

            return _delivery
        return delivery_mechanism

    @pytest.fixture
    def mock_mechanism(self, mock_delivery):
        def mechanism(
            drm_scheme: Optional[str] = None,
            content_type: Optional[str] = "application/epub+zip",
        ) -> LicensePoolDeliveryMechanism:
            def _mechanism_eq(self, other):
                return self.delivery_mechanism == other.delivery_mechanism

            def _mechanism_repr(self):
                return f"LicensePoolDeliveryMechanism(delivery_mechanism={self.delivery_mechanism})"

            _mechanism = MagicMock(spec=LicensePoolDeliveryMechanism)
            _mechanism.delivery_mechanism = mock_delivery(drm_scheme, content_type)
            _mechanism.__eq__ = _mechanism_eq
            _mechanism.__repr__ = _mechanism_repr
            return _mechanism
        return mechanism

    @pytest.fixture
    def sample_data_0(self, mock_mechanism):
        """An arrangement of delivery mechanisms taken from a working database."""
        return [
            mock_mechanism("application/vnd.adobe.adept+xml", "application/epub+zip"),
            mock_mechanism(None, "application/audiobook+json"),
        ]
```

Reply:
Ah, I see, like a general purpose fixture that any test can reuse?

No, the `_show` function isn't used right now, but it is something you end up wanting very quickly should a test fail, so I left it in.

Reply:
For now just creating fixtures as we are adding / modifying tests, and generalizing them as we see common need seems like a way to slowly move in that direction.

Reply:
I added the fixtures you gave above almost verbatim, and adjusted the tests to use them. Seems to work fine!